### PR TITLE
[WIP]: add password field to certmanager

### PR DESCRIPTION
**Pull request by mimugmail**
I thought this might be the way to go but I get some errors when creating a certificate:

```
The following input errors were detected:

openssl library returns: error:0E06D06C:configuration file routines:NCONF_get_string:no value
openssl library returns: error:0E06D06C:configuration file routines:NCONF_get_string:no value
openssl library returns: error:0E06D06C:configuration file routines:NCONF_get_string:no value
openssl library returns: error:0E06D06C:configuration file routines:NCONF_get_string:no value
```

I'm quite sure it's badly escaped or probably it's not the way to go, just an idea.

### CODE REVIEW

**Comment by fichtner** [comment]
is this for exporting ?

**Comment by mimugmail** [comment]
Yes, the PHP function defaults to null, thought this might work. My first journey to legacy code :)

**Comment by fichtner** [comment]
the biggest issue here is the p12 export is a small button and then you'd have to type the password *twice* in order to reliably export a protected file so that requires a modal and further tweaking for the page POST behaviour...

**Comment by fichtner** [comment]
You want to store the export password in the config? Per certificate or globally? It would be better to have a password prompt at export time. That’s why you want the input twice to verify the password is correct. See config export with password...

**Comment by fabianfrz** [comment]
@fichtner or use a JS PRNG to generate a password in the web interface and print it in a modal afterwards - should work too.

**Comment by mimugmail** [comment]
@fichtner at first I need a little success to motivate myself working with legacy stuff. :)
I'll look at the config export part, thx.

**Comment by mimugmail** [comment]
Ok, now you have a Password field when creating a new certificate. It's stored in config.xml and when exporting p12 it'll take the one from config and set it correctly :)

File: src/www/system_certmanager.php [hunk 1]
```diff
@@ -421,6 +420,8 @@ function () {
 
                 $cert['descr'] = $pconfig['descr'];
 
+                $cert['password'] = $pconfig['password'];
+
```
Comment (fabianfrz) [inline src/www/system_certmanager.php]: you can also use an @ to suppress the warning if possible.

**Comment by fichtner** [comment]
Okay, I like the idea to add a password input on the certificate details page, but storing it in the config is probably counter-productive in the long run. Why not remove the p12 export button from the list and add a small extra field for password input and download?

**Comment by fabianfrz** [comment]
@fichtner I would use the bootstrap dialog for that. Just for linking the button to another JS function.

**Comment by fichtner** [comment]
@fabianfrz yes, what I said here https://github.com/opnsense/core/pull/2926#issuecomment-439620550 but it's likely tricky for @mimugmail to pull this off just yet

**Comment by fabianfrz** [comment]
@mimugmail can you try this for the dialog solution?

```diff
diff --git a/src/www/system_certmanager.php b/src/www/system_certmanager.php
index 43c06e1e2..50847680e 100644
--- a/src/www/system_certmanager.php
+++ b/src/www/system_certmanager.php
@@ -559,6 +559,39 @@ if (empty($act)) {
       });
     });
 
+    $('.p12btn').on('click', function(event) {
+        event.preventDefault();
+        var link = $(this).attr('href');
+        BootstrapDialog.show({
+            type:BootstrapDialog.TYPE_INFO,
+            title: "<?= gettext("Certificates");?>",
+            message: "<?=sprintf(gettext('If you want to protect the file with a password, please enter it here (leave empty for no password): %s'), '<input type=\\"text\\" name=\\"p12pw\\" id=\\"p12pw\\" />')?>",
+            buttons: [
+                {
+                    label: "<?=gettext("Close");?>",
+                    action: function(dialogRef) {
+                        dialogRef.close();
+                    }
+                }, {
+                    label: "<?=gettext("Download");?>",
+                    action: function(dialogRef) {
+                        dialogRef.close();
+                        $.post(link, "password=" + encodeURI($('#p12pw').val()) ,function (result)
+                        {
+                            var tmp=document.createElement('a');
+                            tmp.href=window.URL.createObjectURL(new Blob([result]));
+                            tmp.download="certificate.p12";
+                            tmp.style.display = 'none';
+                            document.body.append(tmp);
+                            tmp.click();
+                            tmp.remove();
+                        });
+                    }
+                }
+            ]
+        });
+    });
+
     $(".act_info").click(function(event){
         event.preventDefault();
         var id = $(this).data('id');
@@ -1305,7 +1338,7 @@ $( document ).ready(function() {
                     <i class="fa fa-download fa-fw"></i>
                   </a>
 
-                  <a href="system_certmanager.php?act=p12&amp;id=<?=$i;?>" class="btn btn-default btn-xs" data-toggle="tooltip" title="<?=gettext("export ca+user cert+user key in .p12 format");?>">
+                  <a href="system_certmanager.php?act=p12&amp;id=<?=$i;?>" class="btn btn-default btn-xs p12btn" data-toggle="tooltip" title="<?=gettext("export ca+user cert+user key in .p12 format");?>">
                       <i class="fa fa-download fa-fw"></i>
                   </a>
 <?php
```

**Comment by fabianfrz** [comment]
@mimugmail FYI: this does not include the backend code.

**Comment by mimugmail** [comment]
@fichtner tbh the only reason for static password is that this is the only way how I'd solve this with my little knowledge :) I'm using here a dedicated OPNsense instance just for PKI stuff and p12 password export is one of the things where I have to move it to a linux machine, extract and make a p12 again :) This could also just be treated as a PoC code to convince you this is a good feature to add. 

@fabianfrz seems to work, but I have no idea how to go on with this. For me it's totally fine when we close this one and go on with your approach :)

**Comment by mimugmail** [comment]
Actually I have many open issues in plugins and no idea of Bootstrap code. If you think it's only a couple of minutes for you and you can afford them, I'd be more than happy to close my PR :)

**Comment by AdSchellevis** [comment]
@mimugmail just leave it open, I'll look at the bootstrap code for you/

**Comment by fabianfrz** [comment]
@AdSchellevis All the frontend code, that is required is already in my diff (just needs to be applied). It is only the backend code for handling the post request missing.

**Comment by AdSchellevis** [comment]
@fabianfrz I'll glue it all together :)

**Comment by AdSchellevis** [comment]
@mimugmail @fabianfrz here you go https://github.com/opnsense/core/commit/8a4c4614420f332222d6fbf006c17bd6a35f7765 can you test and confirm?

**Comment by fabianfrz** [comment]
works

**Comment by mimugmail** [comment]
yep, really nice :) Thanks! Also thanks to @fabianfrz for taking the time!

**Pull request closed without merging**